### ISSUE TITLE
2.0 Top Crasher: webView is null in onDestroy

### ISSUE THREAD

**Issue by Sdaswani** [mozilla-mobile/firefox-tv#694]
*mcomella edit*: Historical bug connections:
- #462: I refactored WebFragment -> IWebViewLifecycleFragment. This moved WebView.destroy to onDestroyView (to match when it's created in the layout). This caused crashes...
- #587: the crash bug.
  - First, we added a null check to `onDestroyView` but this still crashed
  - Then we moved webView destruction back to `onDestroy`, without the null check that was previously there from Focus.

---

This is happening more than the Parcel one - probably warrants a service release.

```
java.lang.RuntimeException
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2367)
    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2435)
    at android.app.ActivityThread.access$800(ActivityThread.java:160)
    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321)
    at android.os.Handler.dispatchMessage(Handler.java:102)
    at android.os.Looper.loop(Looper.java:135)
    at android.app.ActivityThread.main(ActivityThread.java:5483)
    at java.lang.reflect.Method.invoke(Native Method)
    at java.lang.reflect.Method.invoke(Method.java:372)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:984)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
Caused by: kotlin.KotlinNullPointerException
    at org.mozilla.focus.web.IWebViewLifecycleFragment.onDestroy(IWebViewLifecycleFragment.kt:82)
    at android.support.v4.app.Fragment.performDestroy(Fragment.java:2609)
    at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1557)
    at android.support.v4.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1740)
    at android.support.v4.app.BackStackRecord.executePopOps(BackStackRecord.java:853)
    at android.support.v4.app.FragmentManagerImpl.executeOps(FragmentManager.java:2577)
    at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2367)
    at android.support.v4.app.FragmentManagerImpl.removeRedundantOperationsAndExecute(FragmentManager.java:2322)
    at android.support.v4.app.FragmentManagerImpl.popBackStackImmediate(FragmentManager.java:851)
    at android.support.v4.app.FragmentManagerImpl.popBackStackImmediate(FragmentManager.java:805)
    at org.mozilla.focus.fragment.ScreenController.showHomeScreen(ScreenController.kt:93)
    at org.mozilla.focus.activity.MainActivity$onCreate$1.onValueChanged(MainActivity.kt:60)
    at org.mozilla.focus.activity.MainActivity$onCreate$1.onValueChanged(MainActivity.kt:55)
    at org.mozilla.focus.architecture.NonNullObserver.onChanged(NonNullObserver.kt:14)
    at android.arch.lifecycle.LiveData.considerNotify(LiveData.java:131)
    at android.arch.lifecycle.LiveData.dispatchingValue(LiveData.java:143)
    at android.arch.lifecycle.LiveData.access$400(LiveData.java:63)
    at android.arch.lifecycle.LiveData$LifecycleBoundObserver.activeStateChanged(LiveData.java:395)
    at android.arch.lifecycle.LiveData$LifecycleBoundObserver.onStateChanged(LiveData.java:378)
    at android.arch.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:353)
    at android.arch.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:291)
    at android.arch.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:331)
    at android.arch.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:137)
    at android.arch.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:123)
    at android.arch.lifecycle.ReportFragment.dispatch(ReportFragment.java:121)
    at android.arch.lifecycle.ReportFragment.onStart(ReportFragment.java:81)
    at android.app.Fragment.performStart(Fragment.java:2077)
    at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:922)
    at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1067)
    at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1049)
    at android.app.FragmentManagerImpl.dispatchStart(FragmentManager.java:1874)
    at android.app.Activity.performStart(Activity.java:6048)
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2330)
    ... 10 more
```

**Comment by Sdaswani** [comment]
![85414f98-f517-4356-bc14-c1c77c0d8326](https://user-images.githubusercontent.com/4879954/37597680-535c7cb8-2b3d-11e8-889a-fe2c650ab32a.png)

**Comment by mcomella** [comment]
The problem is that webView is null going into `onDestroy`. I think either:
- onDestroy is getting called twice (which seems to happen sometimes though it's not well documented. Perhaps https://stackoverflow.com/a/23664121 explains it?)
- onDestroy is getting called with onCreateView being called. This doesn't follow [the lifecycle](https://cdn-images-1.medium.com/max/800/0*YWCr1M1pwWfvCzvZ.png) but note that the lifecycle is a third-party creation.

**Comment by mcomella** [comment]
stack trace highlights:
```
Caused by: kotlin.KotlinNullPointerException
    at org.mozilla.focus.web.IWebViewLifecycleFragment.onDestroy(IWebViewLifecycleFragment.kt:82)
...
    at android.support.v4.app.FragmentManagerImpl.popBackStackImmediate(FragmentManager.java:805)
    at org.mozilla.focus.fragment.ScreenController.showHomeScreen(ScreenController.kt:93)
    at org.mozilla.focus.activity.MainActivity$onCreate$1.onValueChanged(MainActivity.kt:60)
...
    at android.app.Activity.performStart(Activity.java:6048)
    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2330)
```

Key facts:
- Some Activity is being created (performLaunchActivity calls onCreate, then onStart which is where we are failing)
- MainActivity.onCreate must have been called (to assign the onValueChanged callback, which is lifecycle aware)
- IWebViewLifecycleFragment...
  - Already exists in the fragment stack (otherwise `popBackStackImmediate` wouldn't try to pop it)
  - Is getting `onDestroy` called while webView is null

Note: [focus does a null check](https://github.com/mozilla-mobile/focus-android/blob/a2f56d92c091ae9f7e492e75fd4d9fde6bbb0a3b/app/src/main/java/org/mozilla/focus/fragment/WebFragment.java#L122) on WebView before calling the destroy methods, while we don't.

**Comment by mcomella** [comment]
From [the docs](https://developer.android.com/guide/components/activities/activity-lifecycle.html#asem): "The system never kills an activity directly. Instead, it kills the process in which the activity runs, destroying not only the activity but everything else running in the process, as well." When it restores such an Activity, it will restore the UI hierarchy state (by default) and allow us to store any data there (which we don't).

Since it kills the process, any memory we stored will die with it, i.e. we'll lose the active session.

StackOveflow's theory (modifying fragment state during onCreate with a non-null savedInstanceState can cause unexpected fragment behavior) is plausible: our process gets killed so the browsing session is still in the hierarchy but onDestroy is called, we recreate MainActivity, and then we pop the fragment state which forces onDestroy to be called for a second time.

**Comment by mcomella** [comment]
If the theory is correct, we can:
1) Continue with the current behavior and check for null in onDestroy, which may break the fragment lifecycle
2) Only apply fragment transactions with savedInstanceState != null, attempt to recover our session with onSaveInstanceState and jump right back to BrowserFragment, which I think is what Android intends to happen. However, Focus removes the state for privacy reasons so this code path is untested
3) Only apply fragment transactions with savedInstanceState != null, but wait until a later BrowserFragment lifecycle method to move back to the homescreen.

2 seems most correct so I'll look into that. To do so, I think I'll try reproducing with the "Don't keep activities" developer option.

---

That being said, I think **this is a user-facing startup crash** (one-time? I'm not sure if the system drops the bundles for a startup crash) so it's high priority.

**Comment by mcomella** [comment]
I can behavior that onDestroy is called twice but I cannot reproduce the crash:
- Add debug log statements to lifecycle methods in Fragment
- Using Pixel 2 emulator, open firefox
- Go to website (I use GVideo)
- CLick home
- Kill the process (I do this via Settings -> Developer -> Running Services -> Cached Processes -> Kill Firefox)
- Open Firefox from the recent app switcher (note: the home screen button will probably not try to launch state)

Expected: BrowserFragment is not called at all (since there is no session)
Actual: BrowserFragment is created and destroyed. The crash does not reproduce because the view hierarchy is created (**perhaps this is dependent on Android version, or in our case the Fire TV's custom code**)

Don't keep activities didn't work: it doesn't actually kill the process, it just destroys the activities (more reason to use objects that have lifecycles scoped by activities, I guess).

**Comment by mcomella** [comment]
I added some historical context in comment 0: https://github.com/mozilla-mobile/firefox-tv/issues/694#issue-306457210

**Comment by mcomella** [comment]
Given the context:
- We (and Focus) didn't crash before the refactor
- This crash seems predictably related to how we restore from the Android lifecycle (code we haven't really changed).
- ^ I can reproduce *similar* behavior

I think a safe fix here would be to just add a null check, which should be safe enough for a re-release.

For correctness, I think we still might want to:
1) Make the webView outlive the fragment lifecycle #654 
2) Save instance state and don't perform fragment transactions if it's non-null, as Android expects.

But that can wait for 2.1.

**Comment by Sdaswani** [comment]
Thanks @mcomella happy to do a re-release, feel free to make a new branch off of releases/v2.0.0.0 and I can push it tonight.

**Comment by Sdaswani** [comment]
So AFAICT from the data we have for 2.0.1, this patch fixed the crash on 7.1 devices, but it's still crashing on 5.1.1 devices. Any reason you can think of for that @mcomella ?

**Comment by mcomella** [comment]
I'm not convinced the crash rate has changed at all – 7.1 devices had fewer crashes so it's possible no one on those versions has actually crashed yet.

Also, afaict, it's impossible for us to crash on a NPE now:
- We check for null before using the value
- It can only be re-assigned in this file but all assignments run on the UI thread so it shouldn't get reassigned after checking but before using

Also, looking at the stack trace:
`at org.mozilla.focus.web.IWebViewLifecycleFragment.onDestroy(IWebViewLifecycleFragment.kt:82)`

We don't have code on line 82 anymore: https://github.com/mozilla-mobile/firefox-tv/blob/ad805da48a2c6059cb271bd81d0708ed58996fe7/app/src/main/java/org/mozilla/focus/iwebview/IWebViewLifecycleFragment.kt#L80-L90

Some possibilities:
- We built and released the wrong branch
- Amazon's crash reporter doesn't aggregate on exact stack trace matches (such as line numbers) and is providing inaccurate data
- Amazon's interface got confused with 2.0.0.0 and 2.0.1 and released an older version instead

The dashboard indicates the version name is "2.0.1" and is pulling it from the manifest though, which rules out releasing the wrong version (unless there is a UI override).

@Sdaswani I'm not sure what's going on here...

**Comment by Sdaswani** [comment]
@mcomella I don't think I released the wrong version, as I had a clean directory when I built it.

``~\Code\firefox-tv [releases/v2.0.1 ≡]> git status``
``On branch releases/v2.0.1``
``Your branch is up-to-date with 'origin/releases/v2.0.1'.``
``nothing to commit, working tree clean``

That line number thing is very odd, and does point to something being wrong on Amazon's side.

I'll keep an eye on the crashes and also check user reviews, let's move along for now.
